pkg/controlplane/controller/legacytokentracking/controller.go

```diff
@@ -0,0 +1,211 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+)
+
+const (
+	configMapName    = "tracked-since"
+	configMapDataKey = "timestamp"
+	freezePeriod     = 30 * time.Minute
+)
+
+// Controller maintains a bool value configmap `tracked-since` in `kube-system`
+// to indicates if the tracking for legacy tokens is enabled in the cluster.
+// For HA clusters, the configmap will be eventually created after all
+// controller instances have enabled the feature. When disabling this feature,
+// existing configmap will be deleted.
+type Controller struct {
+	configMapClient   corev1client.ConfigMapsGetter
+	namespaceClient   corev1client.NamespacesGetter
+	namespaceExists   bool
+	configMapInformer cache.SharedIndexInformer
+	configMapCache    cache.Indexer
+	configMapSynced   cache.InformerSynced
+	queue             workqueue.RateLimitingInterface
+
+	// enabled controls the behavior of the controller: if enabled is true, the
+	// 'tracked-since' configmap will be created; otherwise, the configmap will
+	// be deleted.
+	enabled bool
+	// freezeUntil records the timestamp until when the controller starts to
+	// create the configmap.
+	freezeUntil time.Time
+}
+
+// NewController returns a Controller struct.
+func NewController(cs kubernetes.Interface) *Controller {
+	informer := corev1informers.NewFilteredConfigMapInformer(cs, metav1.NamespaceSystem, 12*time.Hour, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermEqualSelector("metadata.name", configMapName).String()
+	})
+
+	c := &Controller{
+		configMapClient:   cs.CoreV1(),
+		namespaceClient:   cs.CoreV1(),
+		namespaceExists:   false,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "legacy_token_tracking_controller"),
+		configMapInformer: informer,
+		configMapCache:    informer.GetIndexer(),
+		configMapSynced:   informer.HasSynced,
+		enabled:           utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			configMap := newObj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.freeze()
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
```

Comment:
What adds the key to the queue once the freeze period ends, the config map does not exist and it should be created?

Reply:
> why enqueue if we just froze and won't recreate for 30 minutes... won't this just no-op during the sync?

removed.

> What adds the key to the queue once the freeze period ends, the config map does not exist and it should be created?

e.g. watch events; resync will put events back to the queue

Reply:
I am not an expert on informers so correct me if I am wrong. The situation is this:

1. no CM that would satisfy the informer filter exists and all the previous events are already processed
2. resync happens but since there are no items in the cache, its state matches the state of the cluster

At this point I wouldn't expect any Add/Update/Delete event to be emitted because there wasn't a change to the cache.

What adds the key to the queue at this point?

Reply:
i added polling behavior which should resolve this.